src/y0/algorithm/identify/idcd.py

```diff
@@ -0,0 +1,326 @@
+
+    :param graph: The causal directed mixed graph (may contain cycles).
+    :param C: Target set of variables to identify within district D.
+    :param D: Consolidated district containing C.
+    :param Q_D: Probability distribution over district D.
+    :param _number_recursions: Recursion depth tracker for logging.
+    :returns: Identified causal effect Q[C].
+    :raises ValueError: If preconditions are violated (C ⊆ D ⊆ V).
+    :raises Unidentifiable: If causal effect cannot be identified.
+    """
+    # line 14
+    line_14(graph, C, D, _number_recursions)
+
+    # line 15
+    subgraph_D = graph.subgraph(D)
+    A = line_15(subgraph_D, C, D, _number_recursions)
+
+    # line 16
+    Q_A = line_16(subgraph_D, Q_D, D, A, _number_recursions)
+
+    # lines 17-18
+    if A == C:
+        logger.debug(f"[{_number_recursions}]: Lines 17-18 - SUCCESS (A = C)")
+        return Q_A
+
+    # lines 19-20
+    if A == D:
+        logger.debug(f"[{_number_recursions}]: Lines 19-20 - FAILURE (A = D)")
+        raise Unidentifiable(
+            f"Cannot identify: ancestral closure equals district (A = D = {sorted(A)})"
+        )
+
+    # lines 21-26
+    return lines_21_26(graph, C, A, D, _number_recursions)
+
+
+def line_14(
+    graph: NxMixedGraph,
+    C: set[Variable],
+    D: set[Variable],
+    recursion_level: int,
+) -> None:
+    """Run line 14 of IDCD algorithm.
+
+    Line 14: require C ⊆ D ⊆ V
+
+    Validates IDCD preconditions.
+
+    :param graph: The causal graph.
+    :param C: Target variable set.
+    :param D: District containing C.
+    :param recursion_level: Current recursion depth for logging.
+    :raises ValueError: If any precondition is violated.
+    """
+    V = set(graph.nodes())
+
+    if not C:
+        raise ValueError("Target set C cannot be empty")
+
+    if not D:
+        raise ValueError("District D cannot be empty")
+
+    if not C.issubset(D):
+        raise ValueError(
+            f"C must be subset of D. C={sorted(C)}, D={sorted(D)}, C\\D={sorted(C-D)}"
+        )
+
+    if not D.issubset(V):
+        raise ValueError(
+            f"D must be subset of V. D={sorted(D)}, V={sorted(V)}, D\\V={sorted(D-V)}"
+        )
+
+    logger.debug(
+        f"[{recursion_level}]: Line 14 - Preconditions satisfied: "
+        f"C={sorted(C)}, D={sorted(D)}, |V|={len(V)}"
+    )
+
+
+def line_15(
+    subgraph_D: NxMixedGraph,
+    C: set[Variable],
+    D: set[Variable],
+    recursion_level: int,
+) -> set[Variable]:
+    r"""Run line 15 of IDCD algorithm.
+
+    Line 15: A ← Anc^G[D](C) ∩ D
+
+    Computes the ancestral closure of C within subgraph G[D].
+
+    :param subgraph_D: Subgraph G[D] induced by district D.
+    :param C: Target variable set.
+    :param D: District containing C.
+    :param recursion_level: Current recursion depth for logging.
+    :returns: Ancestral closure A.
+    """
+    A = set().union(*(subgraph_D.ancestors_inclusive(c) for c in C)) & D
```

Comment:
```suggestion
    A = subgraph_D.ancestors_inclusive(C) & D
```

`NxMixedGraph.ancestors_inclusive` accepts a collection of variables or a single variable. When given a collection, it takes the same union as what you wrote